### PR TITLE
feat(dup): add metrics for duplication

### PR DESCRIPTION
**Pull request by neverchanje**
This PR introduces several metrics for duplication:

    perf_counter_wrapper _counter_dup_log_read_bytes_rate;
    perf_counter_wrapper _counter_dup_log_read_mutations_rate;
    perf_counter_wrapper _counter_dup_shipped_bytes_rate;
    perf_counter_wrapper _counter_dup_confirmed_rate;
    perf_counter_wrapper _counter_dup_pending_mutations_count;
    perf_counter_wrapper _counter_dup_time_lag;

## log_read_bytes_rate

**name**: `replica*eon.replica_stub*dup.log_read_bytes_rate`

Calculates the bytes rate read from the private-log. 

The curve line is usually identical with `replica*eon.replica_stub*shared.log.recent.write.size`. Because when everything normal, what is written is what duplicated, then:

log_read_bytes_rate = shared.log.recent.write.size = shipped_bytes_rate

But in some failure conditions, `log_read_bytes_rate` may be much larger, which can be used to identify if log reading during duplication works abnormally.

## log_read_mutations_rate

**name**: `eon.replica_stub dup.log_read_mutations_rate`

Read rate in mutations number. The same usage as "log_read_bytes_rate".

## shipped_bytes_rate

**name**: `eon.replica_stub dup.shipped_bytes_rate`

The network output bytes for successfully delivered duplication_request.

In some failure conditions, the curve may be dropped to 0, for example when the inter-cluster network is unavailable.

## confirmed_rate

- `eon.replica_stub dup.confirmed_rate`

The rate of confirmed writes, which indicates the number of writes that are duplicated and also confirmed by meta server.

## pending_mutations_count

- `eon.replica_stub dup.pending_mutations_count`

The number of writes that are **not** duplicated, this is one of the most important metrics for duplication. The more pending means weaker consistency. By practice, it's recommended to set an alarm threshold for this metric. Beyond the threshold, the duplication should 

## time_lag(ms)

- `eon.replica_stub dup.time_lag(ms)`

The "latency" between 1. time of the client write arrives at replica server 2. time that the write duplicated and applied to the remote cluster.

t0 -> t1 -> t2
client -> replica server -> remote cluster
time_lag = t2-t1

### CODE REVIEW

File: src/dist/replication/common/replication_common.cpp [hunk 1]
```diff
@@ -49,7 +49,7 @@ replication_options::replication_options()
     delay_for_fd_timeout_on_start = false;
     empty_write_disabled = false;
     allow_non_idempotent_write = false;
-    duplication_disabled = true;
+    duplication_disabled = false;
```
Comment (acelyc111) [inline src/dist/replication/common/replication_common.cpp]: 这种开关类的叫xxx_enable更直观吧，这不是一个好传统

File: src/dist/replication/common/replication_common.cpp [hunk 2]
```diff
@@ -273,9 +273,6 @@ void replication_options::initialize()
 
     duplication_disabled = dsn_config_get_value_bool(
         "replication", "duplication_disabled", duplication_disabled, "is duplication disabled");
-    if (allow_non_idempotent_write && !duplication_disabled) {
```
Comment (acelyc111) [inline src/dist/replication/common/replication_common.cpp]: 为什么不要这个约束了？

File: src/dist/replication/lib/duplication/replica_duplicator_manager.cpp [hunk 3]
```diff
@@ -119,5 +119,14 @@ void replica_duplicator_manager::update_confirmed_decree_if_secondary(decree con
     }
 }
 
+int64_t replica_duplicator_manager::get_pending_mutations_count() const
+{
+    int64_t total = 0;
+    for (auto &dup : _duplications) {
```
Comment (acelyc111) [inline src/dist/replication/lib/duplication/replica_duplicator_manager.cpp]: 加const吧

File: src/dist/replication/lib/duplication/replica_duplicator.cpp [hunk 4]
```diff
@@ -160,5 +165,13 @@ decree replica_duplicator::get_max_gced_decree() const
     return _replica->private_log()->max_gced_decree(_replica->get_gpid());
 }
 
+uint64_t replica_duplicator::get_pending_mutations_count() const
+{
+    // it's not atomic to read last_committed_decree in not-REPLICATION thread pool,
+    // but enough for approximate statistic.
+    uint64_t cnt = _replica->last_committed_decree() - progress().last_decree;
+    return cnt > 0 ? cnt : 0;
```
Comment (acelyc111) [inline src/dist/replication/lib/duplication/replica_duplicator.cpp]: 必然是cnt >= 0的吧？

File: src/dist/replication/lib/duplication/replica_duplicator.cpp [hunk 5]
```diff
@@ -138,6 +139,10 @@ error_s replica_duplicator::update_progress(const duplication_progress &p)
                        _progress.last_decree,
                        _progress.confirmed_decree);
     }
+    if (_progress.confirmed_decree > last_confirmed_decree) { // progress advances
```
Comment (levy5307) [inline src/dist/replication/lib/duplication/replica_duplicator.cpp]: 这个注释补充一下？看不懂写的啥

File: src/dist/replication/lib/duplication/replica_duplicator.cpp (hunk 4 above)
Comment (levy5307) [inline src/dist/replication/lib/duplication/replica_duplicator.cpp]: 用max(cnt, 0)好一些吧

File: src/dist/replication/lib/replica_stub.cpp [hunk 6]
```diff
@@ -219,6 +219,38 @@ void replica_stub::install_perf_counters()
...
+        COUNTER_TYPE_VOLATILE_NUMBER,
+        "number of mutations pending for duplication");
+    _counter_dup_time_lag.init_app_counter(
+        "eon.replica_stub",
+        "dup.time_lag(ms)",
+        COUNTER_TYPE_NUMBER_PERCENTILES,
+        "time (in ms) lag between master and slave in the duplication");
+
+    // <- Cold Backup Metrics ->
+
```
Comment (levy5307) [inline src/dist/replication/lib/replica_stub.cpp]: 这个空行去掉吧

File: src/dist/replication/common/replication_common.cpp (hunk 2 above)
Comment (neverchanje) [reply src/dist/replication/common/replication_common.cpp]: 暂时不考虑这个约束，一方面因为我们线上默认开启 `allow_non_idempotent_write`，一方面是开启热备份的表可以在接入层面对业务进行要求，不一定要写死在程序里。

File: src/dist/replication/lib/replica_stub.cpp (hunk 6 above)
Comment (neverchanje) [reply src/dist/replication/lib/replica_stub.cpp]: 和 `<- Duplication Metrics ->` 下面的空行保持一致

File: src/dist/replication/lib/duplication/replica_duplicator.cpp (hunk 4 above)
Comment (neverchanje) [reply src/dist/replication/lib/duplication/replica_duplicator.cpp]: 不一定，这里加了注释，last_committed_decree 多线程访问可能会访问到旧值，差值甚至会小，以前测试到这个问题，这里才这么写。

File: src/dist/replication/common/replication_common.cpp (hunk 2 above)
Comment (hycdong) [reply src/dist/replication/common/replication_common.cpp]: 那现在开热备的表能够同时进行非幂等吗？

Comment (neverchanje) [reply src/dist/replication/common/replication_common.cpp]: 目前不行，如果禁止的话，可能也不会依赖配置来禁止非幂等的写，毕竟一个集群可能有的表热备份，有的表不热备份。

File: src/dist/replication/lib/duplication/replica_duplicator.cpp (hunk 4 above)
Comment (neverchanje) [reply src/dist/replication/lib/duplication/replica_duplicator.cpp]: 改成了 `return cnt > 0 ? static_cast<uint64_t>(cnt) : 0;`
考虑到涉及到一个类型转换，这么写比较好写。用 max 收益不大。

File: src/dist/replication/common/replication_common.cpp (hunk 2 above)
Comment (hycdong) [reply src/dist/replication/common/replication_common.cpp]: 嗯，好，那目前有什么措施保证在热备的表没有进行非幂等操作呢？如果这个由业务控制而我们代码上没有限制，感觉还是有些不安全

Comment (neverchanje) [reply src/dist/replication/common/replication_common.cpp]: 目前还没有，初步的想法是在 pegasus 那边改，遇到 INCR 和 CHECK_AND_SET 就写一个 empty write，然后返回错误。但是还没实现。HBase 是支持热备份 INCR 的，就是复制的过程中，把 INCR 转为 PUT，但是这个流程 pegasus 这边很难写。

Comment (hycdong) [reply src/dist/replication/common/replication_common.cpp]: 嗯，好，记个TODO吧，这个最好还是代码上限制一下，靠业务的自觉性太不安全了

**Pull request merged**